feat: add MutatingPolicy proposal

realshuting opened the pull request.

MariamFahmy98 commented on proposals/mutatingpolicies.md:
```diff
@@ -0,0 +1,159 @@
+      operations:  [CREATE]
+      resources:   ["configmaps"]
+  ## custom targets filters
+  targetMatchConstraints:
+  - # namespaceSelector:
+    # objectSelector:
+    # excludeResourceRules:
+    resourceRules:
+    - apiGroups:   [""]
+      apiVersions: [v1]
+      resources:   ["secrets"]
```
I assume we'll be using the `MatchConstraints` struct from VAPs/MAPs. The `operations` field is present, but it doesn't seem relevant for our use case. Since we won’t be generating MAPs for mutate existing rules, should we define our own struct to represent the target resources instead?

MariamFahmy98 commented on proposals/mutatingpolicies.md:
````diff
@@ -0,0 +1,159 @@
+The built-in variable `target` is supported to reference the target object. For example, the variable `target` is used below to filter secret's namespace via `targets.matchConditions`, and to reference `.spec` object of the secret.
+
+```yaml
+  targetsMatchConstraints:
+  - resourceRules:
+    - apiGroups:   [""]
+      apiVersions: [v1]
+      resources:   ["secrets"]
+  targetMatchConditions:
+  - name: "skip kube-system namespace"
+    expression: >-
+      target.metadata.namespace != "kube-system"
````
This will return a list of secrets, meaning we'll have multiple targets. Are we planning to loop through each target and apply the mutation individually?

realshuting replied:
Yes, mutations are applied to each target resource individually.

We can have another `mutations.foreach` field to handle loop iterations.

JimBugwadia commented on proposals/mutatingpolicies.md:
```diff
@@ -0,0 +1,159 @@
+metadata:
+  name: add-label-to-existing-secret
+spec:
+  mutateExistingOnPolicyUpdate: true
```
Should we move to a `mutateConfigurations` object?

JimBugwadia replied:
For example:

```sh
k explain ivpol.spec.validationConfigurations
GROUP:      policies.kyverno.io
KIND:       ImageValidatingPolicy
VERSION:    v1alpha1

FIELD: validationConfigurations <Object>


DESCRIPTION:
    ValidationConfigurations defines settings for mutating and verifying image
    digests, and enforcing image verification through signatures.

FIELDS:
  mutateDigest	<boolean>
    MutateDigest enables replacement of image tags with digests.
    Defaults to true.

  required	<boolean>
    Required validates that images are verified, i.e., have passed a signature
    or attestation check.

  verifyDigest	<boolean>
    VerifyDigest validates that images have a digest.
```

realshuting replied:
I have added a new field `spec.evaluation.mutateExisting` to enable mutation of existing resources. The same `spec.evaluation` field is supported in both vpol and ivpol for configurable evaluation options.

```
$ k explain ivpol.spec.evaluation              
KIND:     ImageValidatingPolicy
VERSION:  policies.kyverno.io/v1alpha1

RESOURCE: evaluation <Object>

DESCRIPTION:
     EvaluationConfiguration defines the configuration for the policy
     evaluation.

FIELDS:
   admission    <Object>
     Admission controls policy evaluation during admission.

   background   <Object>
     Background controls policy evaluation during background scan.

   mode <string>
     Mode is the mode of policy evaluation. Allowed values are "Kubernetes" or
     "JSON". Optional. Default value is "Kubernetes".
```

```
$ k explain vpol.spec.evaluation 
KIND:     ValidatingPolicy
VERSION:  policies.kyverno.io/v1alpha1

RESOURCE: evaluation <Object>

DESCRIPTION:
     EvaluationConfiguration defines the configuration for the policy
     evaluation.

FIELDS:
   admission    <Object>
     Admission controls policy evaluation during admission.

   background   <Object>
     Background controls policy evaluation during background scan.

   mode <string>
     Mode is the mode of policy evaluation. Allowed values are "Kubernetes" or
     "JSON". Optional. Default value is "Kubernetes".
```

realshuting commented:
> Looks great! Added minor comment, and also discussed we should add a `mutateExisting` field and make `targetMatchConstraints` optional i.e. by default the matchConstraints should apply to both new and existing resources.

Thanks! I have renamed the field to `mutateExisting`, and changed the `targetMatchConditions` to `conditions`. Now with `conditions` we can support general condition filters, both built-in variables `target` and `object` can be used in this field for CEL-based condition matching.
